src/main/java/enmasse/kafka/bridge/SinkBridgeEndpoint.java

```diff
@@ -473,4 +388,244 @@ private ErrorCondition checkFilters(Object partition, Object offset) {
+			if (subscribeResult.failed()) {
+				LOG.error("Error getting partition info {}", kafkaTopic, subscribeResult.cause());
+				String message = subscribeResult.cause().getMessage();
+				ErrorCondition condition =
+						new ErrorCondition(Symbol.getSymbol(Bridge.AMQP_ERROR_KAFKA_SUBSCRIBE),
+								"Error subscribing" + (message != null ? ": " + message : ""));
+				sendProtonError(condition);
+				return;
+			}
+			this.consumer.partitionsRevokedHandler(partitions -> {
+				
+				LOG.debug("Partitions revoked {}", partitions.size());
+				
+				if (!partitions.isEmpty()) {
+					
+					if (LOG.isDebugEnabled()) {
+						for (TopicPartition partition : partitions) {
+							LOG.debug("topic {} partition {}", partition.getTopic(), partition.getPartition());
+						}
+					}
+				
+					// Sender QoS unsettled (AT_LEAST_ONCE), need to commit offsets before partitions are revoked
+					
+					if (this.qos == ProtonQoS.AT_LEAST_ONCE) {
+						// commit all tracked offsets for partitions
+						SinkBridgeEndpoint.this.commitOffsets(true);
+					}
+				}
+			});
+			
+			this.consumer.partitionsAssignedHandler(partitions -> {
+				LOG.debug("Partitions assigned {}", partitions.size());
+				if (!partitions.isEmpty()) {
+					if (LOG.isDebugEnabled()) {
+						for (TopicPartition partition : partitions) {
+							LOG.debug("topic {} partition {}", partition.getTopic(), partition.getPartition());
+						}
+					}
+					partitionsAssigned();
+				} else {
+					ErrorCondition condition =
+							new ErrorCondition(Symbol.getSymbol(Bridge.AMQP_ERROR_NO_PARTITIONS),
+									"All partitions already have a receiver");
+					sendProtonError(condition);
+				}
+			});
+		});
+	}
+	
+	/**
+	 * Callback to process a kafka record
+	 * @param record The record
+	 */
+	private void handleKafkaRecord(KafkaConsumerRecord<K, V> record) {
+
+		LOG.debug("Processing key {} value {} partition {} offset {}", 
+				record.key(), record.value(), record.partition(), record.offset());
+
+		
+		LOG.debug("Fetched {} records [{}]", 1, this.qos);
+		switch (this.qos){
+		case AT_MOST_ONCE:
+			// Sender QoS settled (AT_MOST_ONCE) : commit immediately and start message sending
+			
+			// 1. when start of batch: immediate commit
+			if (startOfBatch()) {
+				LOG.debug("Start of batch in {} mode => commit()", this.qos);
+				this.consumer.commit(ar -> {
+					if (ar.failed()) {
+						{
+							LOG.error("Error committing ... {}", ar.cause().getMessage());
+							//sendProtonError(newError(Bridge.AMQP_ERROR_KAFKA_SYNC, "Error in commit", ar.cause()));
+						}
+					}
+				});
+			}
+			// 2. commit succeeded, so we can enqueue record for sending
+			LOG.debug("Received from Kafka partition {} [{}], key = {}, value = {}", record.partition(), record.offset(), record.key(), record.value());
+			
+			// 3. start message sending
+			protonSend(new KafkaMessage<K, V>(record.partition(), record.offset(), record.record()));
```

Comment:
the previous version used the commitSync() on the native Kafka client so we sent the AMQP messages only when commit went well. With this way we are sending the record even if the commit fails (it's execute asynchronously). It means that AT_MOST_ONCE isn't guaranteed. For example ... the commit fails, the message is sent in any case and if the bridge crashes and then restarts, the client reconnects it could receive the previous message twice. Wdyt ?